core/src/main/java/dk/alexandra/fresco/framework/util/HmacDrbg.java

```diff
@@ -0,0 +1,143 @@
+
+import dk.alexandra.fresco.framework.MPCException;
+import java.security.InvalidKeyException;
+import java.security.NoSuchAlgorithmException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.function.Supplier;
+import javax.crypto.Mac;
+import javax.crypto.SecretKey;
+import javax.crypto.spec.SecretKeySpec;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class implements the HMac-DRBG (HMac-Deterministic Random Bit Generator) as specified in
+ * https://csrc.nist.gov/publications/detail/sp/800-90a/rev-1/final
+ *
+ * <p>
+ * Note that this class is not threadsafe.
+ * </p>
+ * <p>
+ * Note also that reseeding (calling {@link #setSeed(byte[])}) should occur at least every 2^42
+ * calls to {@link #nextBytes(byte[])} to be secure. No signal will be send about this.
+ * </p>
+ */
+public class HmacDrbg implements Drbg {
+
+  private final Logger logger = LoggerFactory.getLogger(HmacDrbg.class);
+  static String DEFAULT_ALGORITHM = "HmacSHA256";
+  private final String algorithm;
+  private Mac mac = null;
+  private byte[] val = null; // value - internally used
+  private byte[] key = null; // key material
+  long reseedCounter;
+  private List<byte[]> seeds;
+  static final long MAX_RESEED_COUNT = (long) Math.pow(2, 48);
+
+  /**
+   * Start with a seed list and the default algorithm. See
+   * {@link #HmacDeterministicRandomBitGeneratorImpl(String, byte[]...)} for further info.
+   *
+   * @param seeds The seeds used. If empty, the default 0 seed will be used. NB: This should happen
+   *        only during testing as this is highly insecure.
+   * @throws NoSuchAlgorithmException If the default HmacSHA256 hash function is not found on the
+   *         system.
+   */
+  public HmacDrbg(byte[]... seeds) throws NoSuchAlgorithmException {
+    this(null, seeds);
+  }
+
+  /**
+   * Deterministic secure random means that given a seed, it is deterministic what the output
+   * becomes next time. This differs from Java's original SecureRandom in that if you give a seed to
+   * this, it merely adds it to the entropy.
+   *
+   * @param seeds The seeds to use. Often a single seed will be more than enough as you can perform
+   *        the {@link #nextBytes(byte[])} operation 2^48 times before the security guarantee does
+   *        not hold and an exception will be thrown.
+   * @param macSupplier The Mac to be used as the HMac hash function. Default is HMacSHA256. If
+   *        null, this implementation will use the default value.
+   * @throws NoSuchAlgorithmException If the <code>algorithm</code> is not found on the system.
+   */
+  public HmacDrbg(Supplier<Mac> macSupplier, byte[]... seeds) throws NoSuchAlgorithmException {
+    this.seeds = new ArrayList<byte[]>();
+    Collections.addAll(this.seeds, seeds);
+    if (this.seeds.size() < 1) {
+      logger.warn("DRBG initialized with no seeds. Using the very insecure 0 seed. "
```

Comment:
We could also just sample a random seed in this case and have a secure default. Either way, the default behavior when no seed is given should be described in the javadoc above.

Reply:
How would you sample a random seed that all parties agree on which is more secure than just the 0 seed? 
But yes, I will note it in the Javadoc